--- linux-2.6.16.9/include/asm-i386/system.h.orig	2006-04-21 13:01:51.000000000 -0600
+++ linux-2.6.16.9/include/asm-i386/system.h	2006-04-21 13:02:47.000000000 -0600
@@ -83,6 +83,8 @@
 #define savesegment(seg, value) \
 	asm volatile("mov %%" #seg ",%0":"=rm" (value))
 
+#endif  /* __KERNEL__ */
+
 /*
  * Clear and set 'TS' bit respectively
  */
@@ -141,8 +143,6 @@
 	__asm__ __volatile__("movl %0,%%cr4": :"r" (x));
 #define stts() write_cr0(8 | read_cr0())
 
-#endif	/* __KERNEL__ */
-
 #define wbinvd() \
 	__asm__ __volatile__ ("wbinvd": : :"memory");
 
