diff -uNr Crypt-SSLeay-0.51.orig/SSLeay.xs Crypt-SSLeay-0.51/SSLeay.xs
--- Crypt-SSLeay-0.51.orig/SSLeay.xs	2003-05-28 00:55:02.000000000 -0600
+++ Crypt-SSLeay-0.51/SSLeay.xs	2006-02-21 23:30:55.000000000 -0700
@@ -105,6 +105,8 @@
 	static int bNotFirstTime;
 	char buf[1024];
         int rand_bytes_read;
+	
+	SSL_library_init();
 
 	if(!bNotFirstTime) {
                SSLeay_add_all_algorithms();
diff -uNr Crypt-SSLeay-0.51.orig/lib/Crypt/SSLeay/CTX.pm Crypt-SSLeay-0.51/lib/Crypt/SSLeay/CTX.pm
--- Crypt-SSLeay-0.51.orig/lib/Crypt/SSLeay/CTX.pm	2001-06-28 13:38:10.000000000 -0600
+++ Crypt-SSLeay-0.51/lib/Crypt/SSLeay/CTX.pm	1969-12-31 17:00:00.000000000 -0700
@@ -1,3 +0,0 @@
-package Crypt::SSLeay::CTX;
-require Crypt::SSLeay;
-1;
