--- mythplugins-0.19/configure.orig	2006-02-12 19:05:31.000000000 -0700
+++ mythplugins-0.19/configure	2006-02-12 19:06:04.000000000 -0700
@@ -452,7 +452,7 @@
     hasit="no"
     if has_library libkhtml ; then
         if has_header kde/khtml_part.h                  || \
-           has_header /opt/kde3/include/khtml_part.h    || \
+           has_header /opt/kde/include/khtml_part.h     || \
            has_header /usr/kde/3.3/include/khtml_part.h || \
            has_header /usr/kde/3.4/include/khtml_part.h || \
            has_header khtml_part.h ; then
--- mythplugins-0.19/settings.pro.orig	2006-02-12 19:14:38.000000000 -0700
+++ mythplugins-0.19/settings.pro	2006-02-12 19:15:12.000000000 -0700
@@ -5,9 +5,9 @@
 INCLUDEPATH += $${PREFIX}/include
 #INCLUDEPATH += /usr/include/cdda
 #INCLUDEPATH += /usr/include/FLAC
-INCLUDEPATH += /usr/kde/3.3/include
+INCLUDEPATH += /opt/kde/include
 
-LIBPATH += /usr/kde/3.3/lib
+LIBPATH += /opt/kde/lib
 LIBS *= -L$${PREFIX}/lib
 
 DEFINES += _GNU_SOURCE
