--- ./mythbrowser/mythbookmarkmanager/mythbookmarkmanager.pro.orig	2005-05-29 02:39:31.000000000 -0600
+++ ./mythbrowser/mythbookmarkmanager/mythbookmarkmanager.pro	2005-05-29 02:39:48.000000000 -0600
@@ -2,6 +2,7 @@
 
 INCLUDEPATH += /usr/include/kde
 INCLUDEPATH += /opt/kde3/include
+INCLUDEPATH += /opt/kde/include
 
 TEMPLATE = lib
 CONFIG += plugin thread warn_on debug
--- mythbrowser/mythbrowser/mythbrowser.pro.orig	2005-05-29 02:46:22.000000000 -0600
+++ mythbrowser/mythbrowser/mythbrowser.pro	2005-05-29 02:46:47.000000000 -0600
@@ -1,7 +1,6 @@
 include (../../settings.pro )
 
-INCLUDEPATH += /usr/include/kde
-INCLUDEPATH += /opt/kde3/include
+INCLUDEPATH += /opt/kde/include
 
 LIBS += -L/usr/local/lib
 
@@ -17,7 +16,7 @@
 INSTALLS += installimages
 
 LIBS += -lmyth-$$LIBVERSION $$EXTRA_LIBS
-LIBS += -L/opt/kde3/lib -lkhtml
+LIBS += -L/opt/kde/lib -lkhtml
 
 # Input
 HEADERS += webpage.h tabview.h
--- ./settings.pro.orig	2005-05-29 02:51:41.000000000 -0600
+++ ./settings.pro	2005-05-29 02:51:27.000000000 -0600
@@ -2,12 +2,13 @@
 CONFIG += release
 
 isEmpty( PREFIX ) {
-    PREFIX = /usr/local
+    PREFIX = /usr
 }
 
 LIBVERSION = 0.18.1
 
 INCLUDEPATH += $${PREFIX}/include
+INCLUDEPATH += /opt/kde/include
 #INCLUDEPATH += /usr/include/cdda
 #INCLUDEPATH += /usr/include/FLAC
 
@@ -17,10 +18,10 @@
 DEFINES += PREFIX=\"$${PREFIX}\"
 
 # Remove for x86-64
-DEFINES += HAVE_MMX
+#DEFINES += HAVE_MMX
 
 release {
-    QMAKE_CXXFLAGS_RELEASE = -O3 -march=pentiumpro -fomit-frame-pointer
+    QMAKE_CXXFLAGS_RELEASE = -march=i486 -mtune=i686 -Os -pipe -fPIE -pie -fstack-protector-all -Wa,--noexecstack -fomit-frame-pointer
     macx {
         # Don't use -O3, it causes some Qt moc methods to go missing
         QMAKE_CXXFLAGS_RELEASE = -O2
