--- linux-2.6.15/include/linux/autoconf.h.orig	2006-01-01 13:13:40.000000000 -0700
+++ linux-2.6.15/include/linux/autoconf.h	2006-01-01 13:14:07.000000000 -0700
@@ -786,7 +786,10 @@
 #define CONFIG_BT_HCIBLUECARD_MODULE 1
 #define CONFIG_BT_HCIBTUART_MODULE 1
 #define CONFIG_BT_HCIVHCI_MODULE 1
-#undef CONFIG_IEEE80211
+/*
+ * Commented out for the external ieee80211
+ * #undef CONFIG_IEEE80211
+ */
 
 /*
  * Device Drivers
