--- ./config/cf/linux.cf.orig	Wed Mar 27 15:00:24 2002
+++ ./config/cf/linux.cf	Wed Mar 27 15:07:22 2002
@@ -86,6 +86,16 @@
 XCOMM libc:	(LinuxCLibMajorVersion./**/LinuxCLibMinorVersion./**/LinuxCLibTeenyVersion)
 XCOMM binutils:	(LinuxBinUtilsMajorVersion)
 
+/* We here at Slackware like to think of ourselves as the "DefaultLinuxDistribution" :) */
+#if LinuxDistribution == DefaultLinuxDistribution
+#define HasZlib			YES
+#endif
+
+/* And this is in case detection ever works */
+#if LinuxDistribution == LinuxSlackware
+#define HasZlib			YES
+#endif
+
 #if LinuxDistribution == LinuxDebian
 # define SystemManDirectory	/usr/share/man
 # define HasPam			YES
