--- ./config/cf/site.def.orig	Sun Jun 25 13:17:29 2000
+++ ./config/cf/site.def	Sun Jun  3 03:03:21 2001
@@ -132,10 +132,11 @@
 */
 
 /*
-#undef DefaultUserPath
-#define DefaultUserPath /bin:/usr/bin:$(BINDIR):/usr/ucb:/usr/local/bin
 */
-
+#undef DefaultUserPath
+#define DefaultUserPath /usr/local/bin:/bin:/usr/bin
+#undef DefaultSystemPath
+#define DefaultSystemPath /sbin:/usr/sbin:/bin:/usr/bin
 
 /* You do NOT need SetUID if you only run the server under xdm */
 /* You MAY need SetUID if users run the server by hand or under xinit */
