--- kdesdk-3.5.2/kioslave/svn/Makefile.am.orig	2006-03-30 00:06:52.000000000 -0700
+++ kdesdk-3.5.2/kioslave/svn/Makefile.am	2006-03-30 00:41:12.000000000 -0700
@@ -6,7 +6,7 @@
 
 kio_svn_la_SOURCES = svn.cpp
 kio_svn_la_LIBADD = -lkio
-kio_svn_la_LDFLAGS = -avoid-version -module $(all_libraries) $(KDE_PLUGIN) $(SVNLD) $(SVN_LIB)
+kio_svn_la_LDFLAGS = -avoid-version -module $(all_libraries) $(KDE_PLUGIN) $(SVNLD) $(SVN_LIB) -lsvn_ra-1 -lsvn_subr-1
 
 protocol_DATA = svn+http.protocol svn+https.protocol svn+file.protocol svn+ssh.protocol svn.protocol
 protocoldir = $(kde_servicesdir)
