diff -uNr freenx-0.5.0-test-2006-03-11-7.orig/nxloadconfig freenx-0.5.0-test-2006-03-11-7/nxloadconfig
--- freenx-0.5.0-test-2006-03-11-7.orig/nxloadconfig	2006-03-11 04:50:22.000000000 -0700
+++ freenx-0.5.0-test-2006-03-11-7/nxloadconfig	2006-05-19 13:24:08.000000000 -0600
@@ -56,7 +56,7 @@
 NX_LICENSE="OS (GPL)"
 
 # Where can different nx components be found
-NX_DIR=/usr
+NX_DIR=/opt/NX
 PATH_BIN=$NX_DIR/bin # if you change that, be sure to also change the public keys
 PATH_LIB=$NX_DIR/lib
 NX_ETC_DIR=/etc/nxserver
@@ -151,7 +151,7 @@
 DEFAULT_X_WM=""
 KILL_DEFAULT_X_WM="1"
 USER_X_STARTUP_SCRIPT=.Xclients
-DEFAULT_X_SESSION=/etc/X11/xdm/Xsession
+DEFAULT_X_SESSION=/usr/lib/X11/xdm/Xsession
 COMMAND_START_KDE=startkde
 COMMAND_START_GNOME=gnome-session
 COMMAND_START_CDE=cdwm
@@ -160,7 +160,7 @@
 [ -z "$COMMAND_XAUTH" ] && COMMAND_XAUTH=/usr/X11R6/bin/xauth
 COMMAND_SMBMOUNT=smbmount
 COMMAND_SMBUMOUNT=smbumount
-COMMAND_NETCAT=netcat
+COMMAND_NETCAT=nc
 COMMAND_SSH=ssh
 COMMAND_SSH_KEYGEN=ssh-keygen
 COMMAND_CUPSD=/usr/sbin/cupsd
@@ -496,7 +496,7 @@
 	[ -z $(echo "$ENABLE_ROOTLESS_MODE" | egrep "^[0|1]$") ] && \
 		ERROR="yes" && echo "Error: Invalid value \"ENABLE_ROOTLESS_MODE=$ENABLE_ROOTLESS_MODE\""
 
-	[ -z "$(strings $PATH_BIN/nxagent | grep 'NXAGENT - Version 1.5.0')"] && \
+	[ -z "$(strings $PATH_BIN/nxagent | grep 'NXAGENT - Version 1.5.0')" ] && \
 		ERROR="yes" && echo "Error: Could not find 1.5.0 version string in nxagent. NX 1.5.0 backend is needed for this version of FreeNX."
 		
 	[ -z $(echo "$ENABLE_USESSION" | egrep "^[0|1]$") ] && \
diff -uNr freenx-0.5.0-test-2006-03-11-7.orig/nxnode freenx-0.5.0-test-2006-03-11-7/nxnode
--- freenx-0.5.0-test-2006-03-11-7.orig/nxnode	2006-03-11 04:50:22.000000000 -0700
+++ freenx-0.5.0-test-2006-03-11-7/nxnode	2006-05-19 13:24:19.000000000 -0600
@@ -136,6 +136,13 @@
 	. /etc/profile
 	[ -f ~/.bash_profile ] && . ~/.bash_profile
 
+	if [ -r /etc/xprofile ]; then
+		source /etc/xprofile
+	fi
+	if [ -r ~/.xprofile ]; then
+		source ~/.xprofile
+	fi
+
 	mkdir -p $USER_FAKE_HOME/.nx/C-$sess_id/pids/
 
 	STARTX=""
