--- bash_completion/bash_completion.orig	2006-02-15 23:25:23.000000000 -0700
+++ bash_completion/bash_completion	2006-02-15 23:26:10.000000000 -0700
@@ -2474,7 +2474,7 @@
 	    fi
 	    # append any available aliases from config files
 	    if [ ${#config[@]} -gt 0 ] && [ -n "$aliases" ]; then
-		hosts=$( compgen -W "$( sed -ne "s/^[Hh][Oo][Ss][Tt]["$'\t '"]*\([^*?]*\)$/\1/p" ${config[@]} )" -- $ocur )
+		hosts=$( compgen -W "$( sed -ne 's/^[Hh][Oo][Ss][Tt][$\t ]*\([^*?]*\)$/\1/p' ${config[@]} )" -- $ocur )
 		COMPREPLY=( ${COMPREPLY[@]} $hosts )
 	    fi
 
